### DISCUSSION TITLE
How to sign in a user without a password?

### DISCUSSION THREAD

**Discussion by notramo** [martenframework/marten#183] [Q&A] [answered]
When using email magic links, TOTP code, or SMS code, or any method where the user does not provide a password, how to sign in the user using `MartenAuth::User`?

**Comment by notramo** [comment]
@ellmetha, is it possible with the current implementation?

**Comment by ellmetha** [comment, marked as answer]
@notramo Yes, it is possible with the current implementation and [marten-auth](https://github.com/martenframework/marten-auth) allows you to do that easily. You just need to call the `User#set_unusable_password` method when creating your user record and then call `MartenAuth#sign_in` with your `User` record and the current request object at sign in time.

For example:

```crystal
# At sign up time.
user = User.new(...)
user.set_unusable_password
user.save!

# At sign in time.
MartenAuth.sign_in(user, request)
```

This technique is used by other shards that provide passwordless authentication, such as [`msa7/multi_auth`](https://github.com/msa7/multi_auth). You can check out [the Marten example]() in this repository for a more concrete example.

**Reply by notramo** [reply]
Thanks. Is it included in the documentation? I could not find it.

**Reply by ellmetha** [reply]
`User#sign_in` is [definitely documented](https://martenframework.com/docs/authentication/introduction#signing-in-users), but we may be missing a mention to the `User#set_unusable_password` method.

Edit: added in 3cf33e1.
